### PR TITLE
Annotate SV calls with overlapping external CNV calls

### PR DESCRIPTION
**Pull request by cwhelan**
This PR annotates calls made by GATK-SV with any overlapping calls from the external CNV call set passed in.

The input CNV call set must be a single-sample VCF with a sample name that matches the sample name from the input reads and uses VCF-spec FORMAT tags to describe copy number (CN,CNQ).

### CODE REVIEW

File: src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/StructuralVariationDiscoveryPipelineSparkIntegrationTest.java [hunk 1]
```diff
@@ -74,8 +79,15 @@ public String toString() {
         final File tempDirNew = BaseTest.createTempDir("new");
         tempDirNew.deleteOnExit();
         Files.createDirectories(Paths.get(tempDirNew.getAbsolutePath()+"/fastq"));
-        tests.add(new Object[]{new StructuralVariationDiscoveryPipelineSparkIntegrationTest.StructuralVariationDiscoveryPipelineSparkIntegrationTestArgs(SVIntegrationTestDataProvider.TEST_BAM_NEW, SVIntegrationTestDataProvider.KMER_KILL_LIST, SVIntegrationTestDataProvider.ALIGNER_INDEX_IMG, tempDirNew.getAbsolutePath())});
-
+        tests.add(new Object[]{
+                new StructuralVariationDiscoveryPipelineSparkIntegrationTest.StructuralVariationDiscoveryPipelineSparkIntegrationTestArgs(
+                        SVIntegrationTestDataProvider.TEST_BAM_NEW,
```
Comment (SHuang-Broad) [inline src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/StructuralVariationDiscoveryPipelineSparkIntegrationTest.java]: What was the reason the test bam is called "new"? I know we replaced the test bam but the old one is gone right? Can we remove the "new'?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java [hunk 2]
```diff
@@ -91,17 +111,45 @@ protected void runTool( final JavaSparkContext ctx ) {
...
+                        ctx.broadcast(header.getSequenceDictionary()),
+                        vcfOutputFileName,
+                        localLogger,
+                        evidenceLinkTree,
+                        assembledEvidenceResults.getReadMetadata(),
+                        broadcastCNVCalls
                 );
     }
 
+    public static String getSampleId(final SAMFileHeader header) {
```
Comment (SHuang-Broad) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java]: I think this method is better living in a "utils" class, like `SVUtils`?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java [hunk 3]
```diff
@@ -91,17 +111,45 @@ protected void runTool( final JavaSparkContext ctx ) {
         final List<EvidenceTargetLink> evidenceTargetLinks = assembledEvidenceResults.getEvidenceTargetLinks();
         final PairedStrandedIntervalTree<EvidenceTargetLink> evidenceLinkTree = makeEvidenceLinkTree(evidenceTargetLinks);
 
+        final SVIntervalTree<VariantContext> cnvCalls;
```
Comment (SHuang-Broad) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java]: It looks like these lines are highly similar to those in `DiscoverVariantsFromContigAlignmentsSAMSpark.runTool()`. 
Could you refactor them into a method in `DiscoverVariantsFromContigAlignmentsSAMSpark` and try to have the `runTool()` in `DiscoverVariantsFromContigAlignmentsSAMSpark` thin (easier for the pipeline to call into)?

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/AnnotatedVariantProducer.java [hunk 4]
```diff
@@ -71,16 +74,18 @@
      * @param inferredType                      inferred type of variant
      * @param altHaplotypeSeq                   alt haplotype sequence (could be null)
      * @param contigAlignments                  chimeric alignments from contigs used for generating this novel adjacency
-     * @param broadcastReference                broadcasted reference
-     *
-     * @throws IOException                      due to read operations on the reference
+     * @param broadcastReference                broadcast reference
+     * @param broadcastSequenceDictionary       broadcast reference sequence dictionary
+     *@param broadcastCNVCalls  @throws IOException                      due to read operations on the reference
```
Comment (SHuang-Broad) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/AnnotatedVariantProducer.java]: doc is off

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/AnnotatedVariantProducer.java [hunk 5]
```diff
@@ -108,9 +113,31 @@ static VariantContext produceAnnotatedVcFromInferredTypeAndRefLocations(final Si
...
             vcBuilder.attribute(GATKSVVCFConstants.SEQ_ALT_HAPLOTYPE, new String(altHaplotypeSeq));
 
+        if (broadcastCNVCalls != null && end > 0) {
+            annotateWithCNVCalls(refLoc, end, vcBuilder, broadcastSequenceDictionary, broadcastCNVCalls);
+        }
+
         return vcBuilder.make();
     }
 
+    private static void annotateWithCNVCalls(final SimpleInterval refLoc,
```
Comment (SHuang-Broad) [inline src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/AnnotatedVariantProducer.java]: a refactoring suggestion: can you make it return a `String` attribute and not taking (hence modifying) a VCBuilder?

File: src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/StructuralVariationDiscoveryPipelineSparkIntegrationTest.java (hunk 1 above)
Comment (cwhelan) [reply src/test/java/org/broadinstitute/hellbender/tools/spark/sv/integration/StructuralVariationDiscoveryPipelineSparkIntegrationTest.java]: Will change, I think this was just an artifact of when we replaced it.

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java (hunk 2 above)
Comment (cwhelan) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java]: Moved to `SVUtils`

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/AnnotatedVariantProducer.java (hunk 5 above)
Comment (cwhelan) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/AnnotatedVariantProducer.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java (hunk 3 above)
Comment (cwhelan) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java]: It seemed like we were trying to leave operations involving the spark context, like actually broadcasting variables, within the `runTool` methods of the tools involved, which is why I didn't end up doing as you suggested. If I refactor this way I'll have to pass the spark context object in. What if I just make a static method to do the load and broadcast on `DiscoverVariantsFromContigAlignmentsSAMSpark` and call it from both places?

Comment (SHuang-Broad) [reply src/main/java/org/broadinstitute/hellbender/tools/spark/sv/StructuralVariationDiscoveryPipelineSpark.java]: Yep, that would be good.

**Pull request merged**